File: SortedArrayUnknownSize.cpp
// Time Complexity - It is O(logn) since we are finding the target using the Binary Search and in one pass.
// T.C. for finding the range is also log(n) since we are doing 2x at every iteration.
// Space Complexity - It is O(n) since the search depends on the number of elements in the array?
// Problems faced - No!
// It runs on Leetcode!
/**
 * // This is the ArrayReader's API interface.
 * // You should not implement it, or speculate about its implementation
 * class ArrayReader {
 *   public:
 *     int get(int index);
 * };
 */

class Solution {
public:
    int search(const ArrayReader& reader, int target) {
        int low = 0;
        int high = 1;
        int mid;
        
        // find the range in which my target value lies.
        while(reader.get(high) < target){
            low = high;
            high = 2*high;
        }
        
        //apply a binary search on my range
        while(low <= high){
            mid = low + (high - low)/2;
            if(reader.get(mid) == target)
                return mid;
            if(reader.get(mid) > target)
                high = mid - 1;
            else
                low = mid + 1;
        }
        
        return -1;
    }
};

File: SortedArrayUnknownSize.py
# // Time Complexity : O(logn)
# // Space Complexity : O(1)


class Solution:
    def search(self, reader: 'ArrayReader', target: int) -> int:
        low = 0 
        high= 1
        while reader.get(high) < target:
            low = high
            high = 2*high
        return self.binarySearch(reader, target, low, high)

    def binarySearch(self, reader, target, low, high):
        while low<=high:
            mid = low(high-low)//2
            value = reader.get(mid)
            if target == value:
                return mid
            if target > value:
                low = mid +1
            else: 
                high = mid-1
        return -1